src/core/thread/mesh_forwarder.cpp

```diff
@@ -754,163 +754,150 @@ otError MeshForwarder::UpdateIp6Route(Message &aMessage)
+
+    // 1. Choose correct MAC Source Address.
+    GetMacSourceAddress(ip6Header.GetSource(), mMacSource);
+
+    // 2. Choose correct MAC Destination Address.
+    if (netif.GetMle().GetRole() == OT_DEVICE_ROLE_DISABLED ||
+        netif.GetMle().GetRole() == OT_DEVICE_ROLE_DETACHED)
     {
-    case OT_DEVICE_ROLE_DISABLED:
-    case OT_DEVICE_ROLE_DETACHED:
+        // Allow only for link-local unicasts and multicasts.
         if (ip6Header.GetDestination().IsLinkLocal() || ip6Header.GetDestination().IsLinkLocalMulticast())
         {
             GetMacDestinationAddress(ip6Header.GetDestination(), mMacDest);
-            GetMacSourceAddress(ip6Header.GetSource(), mMacSource);
         }
         else
         {
-            ExitNow(error = OT_ERROR_DROP);
+            error = OT_ERROR_DROP;
         }
 
-        break;
+        ExitNow();
+    }
+
+    if (ip6Header.GetDestination().IsMulticast())
+    {
+        mMacDest.mLength = sizeof(mMacDest.mShortAddress);
 
-    case OT_DEVICE_ROLE_CHILD:
-    case OT_DEVICE_ROLE_ROUTER:
-    case OT_DEVICE_ROLE_LEADER:
+        // With the exception of MLE multicasts, a Thread End Device transmits multicasts,
+        // as IEEE 802.15.4 unicasts to its parent.
+        if (netif.GetMle().GetRole() == OT_DEVICE_ROLE_CHILD && !aMessage.IsSubTypeMle())
+        {
+            mMacDest.mShortAddress = netif.GetMle().GetNextHop(Mac::kShortAddrBroadcast);
+        }
+        else
+        {
+            mMacDest.mShortAddress = Mac::kShortAddrBroadcast;
+        }
+    }
+    else if (ip6Header.GetDestination().IsLinkLocal())
+    {
+        GetMacDestinationAddress(ip6Header.GetDestination(), mMacDest);
+    }
+    else
+    {
         if (netif.GetMle().IsMinimalEndDevice())
         {
-            if (aMessage.IsLinkSecurityEnabled())
-            {
-                mMacDest.mLength = sizeof(mMacDest.mShortAddress);
-
-                if (ip6Header.GetDestination().IsLinkLocalMulticast())
-                {
-                    mMacDest.mShortAddress = Mac::kShortAddrBroadcast;
-                }
-                else
-                {
-                    mMacDest.mShortAddress = netif.GetMle().GetNextHop(Mac::kShortAddrBroadcast);
-                }
-
-                GetMacSourceAddress(ip6Header.GetSource(), mMacSource);
-            }
-            else if (ip6Header.GetDestination().IsLinkLocal() || ip6Header.GetDestination().IsLinkLocalMulticast())
-            {
-                GetMacDestinationAddress(ip6Header.GetDestination(), mMacDest);
-                GetMacSourceAddress(ip6Header.GetSource(), mMacSource);
-            }
-            else
-            {
-                ExitNow(error = OT_ERROR_DROP);
-            }
+            mMacDest.mLength       = sizeof(mMacDest.mShortAddress);
+            mMacDest.mShortAddress = netif.GetMle().GetNextHop(Mac::kShortAddrBroadcast);
         }
 
 #if OPENTHREAD_FTD
```

Comment:
Add an `#else` to assert if we are here and `IsMinimalEndDevice()` returns `false`?